### PR TITLE
Display evaluation result overlays at end of sexp

### PR DESCRIPTION
**Pull request by yuhan0**
Rename argument from point -> where to match cider--make-result-overlay and
update docstring
-----------------

Before submitting the PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [x] The commits are consistent with our [contribution guidelines](../.github/CONTRIBUTING.md)
- [ ] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing (`make test`)
- [x] All code passes the linter (`make lint`) which is based on [`elisp-lint`](https://github.com/gonewest818/elisp-lint) and includes
  - [byte-compilation](https://www.gnu.org/software/emacs/manual/html_node/elisp/Byte-Compilation.html), [`checkdoc`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Tips.html), [check-declare](https://www.gnu.org/software/emacs/manual/html_node/elisp/Declaring-Functions.html), packaging metadata, indentation, and trailing whitespace checks.
- [x] You've updated the [changelog](../blob/master/CHANGELOG.md) (if adding/changing user-visible functionality)
- [ ] You've updated the [user manual](../blob/master/doc) (if adding/changing user-visible functionality)

### CODE REVIEW

**Comment by bbatsov** [comment]
Can you add some screenshots before and after?

**Comment by yuhan0** [comment]
Before:
image

After:
image

**Comment by yuhan0** [comment]
I realise a simpler change would be modifying the semantics of `cider--make-result-overlay` to treat a single marker as placing the overlay there instead of the EOL. 

That function itself is rather confusing- the docstrings, comments and code don't match up. 

```
  Place an overlay displaying VALUE at the end of line.
The overlay itself is placed from
beginning to end of current line.
```
...
```
        ;; Make sure the overlay is actually at the end of the sexp.
```
...
```
        (let* (beg (if (consp where)
                        (car where)
                      (save-excursion
                        (clojure-backward-logical-sexp 1)
                        (point))))
               (end (if (consp where)
                        (cdr where)
                      (line-end-position)))

**Comment by yuhan0** [comment]
And a related change would be in `cider--debug-display-result-overlay` which would display intermediate results after the respective forms. 
Making the above semantic change to `cider--make-result-overlay` automatically solves this as well.

**Comment by bbatsov** [comment]
I think I just remembered why we did it the way it was - the overlay moving the code to the right seemed kind of weird and we felt that some people would not like it. I'm not sure how other editors are handling inline evaluations for comparison. Obviously we can always make this configurable.

**Comment by bbatsov** [comment]
//cc @Malabarba

**Comment by yuhan0** [comment]
For what it's worth, VS Code's Calva displays results end-of-sexp:
image

whereas Atom's Proto-repl displays it end-of-line:
image


One small advantage of Calva's approach is that you can see several intermediate results on the same line
image
(I don't think Cider currently supports multiple overlays)

**Comment by yuhan0** [comment]
Another ergonomic issue: long lines cause the overlay to be truncated or pushed off screen completely
image
and you end up having to shift things around or use `cider-inspect-last-result`. 
That's especially troublesome in the debugger where editing the buffer causes you to lose the state, and there's no easy way to invoke the inspector on the current value.

This seems more natural:
image

**Comment by bbatsov** [comment]
> (I don't think Cider currently supports multiple overlays)

No, it doesn't. Your points are good, but knowing how users typically react to this type of changes (usually some of them don't like them) I'd really prefer for us to make this a configuration option instead of just changing the default behaviour. 

Thanks for looking into the behaviour of other editors as well!

**Comment by yuhan0** [comment]
Would you like it to be a separate boolean variable like `cider-display-overlays-at-eol` or expand the options to ﻿`cider-use-overlays`?

**Comment by bbatsov** [comment]
A separate one variable, but not boolean - `cider-overlays-position` (or something like this) with values `'at-point` and `'at-eol`. We might come up with other positions down the road (e.g. above/under) and I want to keep our options open.

**Comment by yuhan0** [comment]
Ok, I named it `cider-result-overlay-position` because `overlays` might be ambiguous.

Should the default remain 'at-eol ? I personally think 'at-point is a more sensible default with the option for the user to revert to the previous behaviour.

**Comment by bbatsov** [comment]
Let's keep the default. I might survey the users to check if we can change it. We should mention this configurable behaviour somewhere in the docs as well.

**Comment by Malabarba** [comment]
I would prefer we keep the default too. 
Your report of overlays getting pushed off the side of the window intrigues me though. I could swear we had some code in there that prints the overlay in a newline (below the current line) when that happens.

File: cider-overlays.el [hunk 1]
```diff
@@ -193,14 +201,19 @@ overlay."
...
         (skip-chars-backward "\r\n[:blank:]")
-        (let* ((beg (if (consp where)
+        (let* ((eolp (eq cider-result-overlay-position 'at-eol))
+               (beg (if (consp where)
                         (car where)
-                      (save-excursion
-                        (clojure-backward-logical-sexp 1)
-                        (point))))
+                      (if eolp
+                          (line-beginning-position)
```
Comment (Malabarba) [inline cider-overlays.el]: Here you're changing the overlay behaviour in the current default.

File: cider-overlays.el [hunk 2]
```diff
@@ -59,17 +59,27 @@ font-locking it."
...
   :group 'cider
   :package-version '(cider . "0.10.0"))
 
+(defcustom cider-result-overlay-position 'at-eol
+  "Where to display result overlays for inline evaluation and the debugger.
+If 'at-eol, display at the end of the line.
+If 'at-point, display at the end of the respective logical sexp."
+  :group 'cider
+  :type ''(choice (const :tag "End of line" at-eol)
+                  (const :tag "End of logical sexp" at-point))
```
Comment (Malabarba) [inline cider-overlays.el]: We can just call this "End of sexp". I think "logical sexp" is more of an internal implementation detail

File: cider-overlays.el (hunk 1 above)
Comment (yuhan0) [reply cider-overlays.el]: Thanks for the review! I was going off the docstring which said "apply the overlay over
  the entire line at that place", but doesn't match up with the implementation. 
I can change it back if that's not appropriate?

Comment (bbatsov) [reply cider-overlays.el]: I think you should change the docstring instead.

File: cider-overlays.el (hunk 2 above)
Comment (bbatsov) [reply cider-overlays.el]: Indeed.

**Comment by yuhan0** [comment]
Ok, made the requested changes and added documentation

**Comment by bbatsov** [comment]
Great work! Thank you! 🙇

File: cider-overlays.el [hunk 3]
```diff
@@ -59,17 +59,27 @@ font-locking it."
...
                  (const :tag "Both" both))
   :group 'cider
   :package-version '(cider . "0.10.0"))
 
+(defcustom cider-result-overlay-position 'at-eol
+  "Where to display result overlays for inline evaluation and the debugger.
+If 'at-eol, display at the end of the line.
+If 'at-point, display at the end of the respective sexp."
+  :group 'cider
+  :type ''(choice (const :tag "End of line" at-eol)
```
Comment (Malabarba) [inline cider-overlays.el]: Sorry for forgetting to finish the review. And thanks for the great work.

Just one thing, I think this line has an extra `'`

**Pull request merged**